### PR TITLE
More realistic gas explosions for bloated and smoker

### PR DESCRIPTION
**Pull request by Coolthulhu**
Addresses the most common cause of #14894

Exploding bloated and smokers will now use a gas propagation function. In open field, it doesn't differ from what happens in master (except that it calculates slower).

When constrained by walls, it will propagate more in the opposite direction.

When constrained by fields of the same type, it will propagate through them, losing some power in the process.

### CODE REVIEW

**Comment by mugling** [comment]
Interestingly I was working on an implementation of this yesterday but you've made much more progress. I was wanting to add an `emit` field to `itype` so we can have proper smoke grenades. How much work would it be to extend your PR to support this as it looks like most of the groundwork is inplace?

**Comment by Coolthulhu** [comment]
Mostly just cleaning up would be required.
If the function is to be used more globally, it needs to be in an accessible file/class, such as `map` or `explosion.cpp`. Having it in `map` would also allow making it less slow (the implementation is rather naive).

**Comment by mugling** [comment]
Do you want to try for that approach - indeed the function you wrote has an identical signature to the `map::emit_field` that I was working on

**Comment by Coolthulhu** [comment]
Moved to field.cpp

Currently doesn't support non-gases properly, because it has a hardcoded check for `TFLAG_PERMEABLE`, but that's easy to fix - just needs the field type to specify whether it is a gas or not.

**Comment by Coolthulhu** [comment]
I added a phase field to fields. Changed the propagation function to respect that by not propagating non-gases through permeable but impassable tiles.

Added propagation to boomer bile on its death as an example.

File: src/enums.h [hunk 1]
```diff
@@ -160,7 +160,7 @@ enum artifact_natural_property {
     ARTPROP_MAX
 };
 
-enum phase_id {
+enum phase_id : int {
```
Comment (mugling) [inline src/enums.h]: `enum class` avoids the need for that ugly untyped `PNULL`

File: src/mondeath.cpp [hunk 2]
```diff
@@ -136,7 +137,7 @@ void mdeath::normal(monster *z)
             }
 
             if( tarp != tripoint_min ) {
-                g->m.spawn_item( tarp, meat );
```
Comment (mugling) [inline src/mondeath.cpp]: Unrelated changes?

Comment (Coolthulhu) [inline src/mondeath.cpp]: Kinda related to critters exploding, though not to propagation
